kore/src/Kore/Strategies/Goal.hs

```diff
@@ -732,19 +742,27 @@ onePathFollowupStep claims axioms =
         , TriviallyValid
         ]
 
+groupStrategy
+    :: [[Rule (AllPathRule Variable)]]
+    -> [Prim (AllPathRule Variable)]
+groupStrategy axiomGroups = do
+    group <- axiomGroups
+    [DerivePar group, Simplify, TriviallyValid]
+
 allPathFirstStep
-    :: [Rule (AllPathRule Variable)]
+    :: [[Rule (AllPathRule Variable)]]
     -> Strategy (Prim (AllPathRule Variable))
-allPathFirstStep axioms =
-    (Strategy.sequence . map Strategy.apply)
+allPathFirstStep axiomGroups =
+    (Strategy.sequence . map Strategy.apply) $
         [ CheckProven
         , CheckGoalStuck
         , CheckGoalRemainder
         , Simplify
         , TriviallyValid
         , RemoveDestination
-        , DerivePar axioms
-        , Simplify
+        ]
+        <> groupStrategy axiomGroups <>
+        [ Simplify
```

Comment:
The groupStrategy function will add `Simplify, TriviallyValid` at the end of each simplification, so I think that most of the time these will be redundant. To cover everything, maybe it would be better to add a special case to `groupStrategy` for an empty input list.

Also, `allPathFollowupStep` seems to go directly to `ResetGoal` after `groupStrategy`, they should do the same thing unless there is a reason to do otherwise.

Reply:
> To cover everything, maybe it would be better to add a special case to `groupStrategy` for an empty input list.

I agree. It looks to me like `groupStrategy` silently does something wrong if it's given an empty list. (If that's not true, then you can ignore everything that follows.) I think we should at least do one of the following:

1. Do the right thing with an empty list. (Best)
1. Take a `NonEmpty` list as its argument.
1. `assert` that its argument is non-empty. (Worst, but still better than silently doing the wrong thing.)